src/event_emitter/mod.rs

```diff
@@ -0,0 +1,82 @@
+// Copyright 2020 Damir Jelić
+// Copyright 2020 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+
+use crate::events::collections::all::{RoomEvent, StateEvent};
+use crate::events::collections::only::Event as NonRoomEvent;
+use crate::events::presence::PresenceEvent;
+use crate::models::Room;
+
+use tokio::sync::Mutex;
+///
+#[async_trait::async_trait]
+pub trait EventEmitter: Send + Sync {
+    // ROOM EVENTS from `IncomingTimeline`
+    /// Fires when `AsyncClient` receives a `RoomEvent::RoomMember` event.
+    async fn on_room_member(&mut self, _: Arc<Mutex<Room>>, _: Arc<Mutex<RoomEvent>>) {}
```

Comment:
This emits a RoomEvent, but shouldn't we emit a [MemberEvent](https://docs.rs/ruma-events/0.18.0/ruma_events/room/member/struct.MemberEvent.html). If we emit a RoomEvent lib users will still need to check the type of the event in the callback which defeats the point of having differing RoomEvent callbacks, no?

Reply:
Yes totally, I thought that would be the way to go too.

Reply:
they should all be that way right?
```rust
pub trait EventEmitter: Send + Sync {
    async fn on_room_member(&mut self, _: Arc<Mutex<Room>>, _: Arc<Mutex<MemberEvent>>) {}
    async fn on_room_name(&mut self, _: Arc<Mutex<Room>>, _: Arc<Mutex<NameEvent>>) {}
    async fn on_room_canonical_alias(&mut self, _: Arc<Mutex<Room>>, _: Arc<Mutex<CanonicalAliasEvent>>) {}
... and so on
```

Reply:
Correct, the lib is doing the de-structuring for the lib user.